src/tools/pub/cmd/pub-outdated.md

```diff
@@ -0,0 +1,251 @@
+   **update the version dependencies** in `pubspec.yaml` and
+   run `pub upgrade`.
+4. **Run `pub outdated`** to confirm that you've updated the dependencies.
+5. **Test** your app or package to confirm that it still works as expected.
+
+
+## Example
+
+You run `pub outdated` in the top directory of your package or app,
+in the same directory as your package or app's
+[`pubspec.yaml`](/tools/pub/pubspec) file.
+Here's an example of running `pub outdated`:
+
+```terminal
+$ pub outdated
+Package              Current   Upgradable  Resolvable  Latest    
+dependencies: all up-to-date
+
+dev_dependencies    
+build_web_compilers  1.2.3     1.2.3       2.10.0      2.10.0    
+
+transitive dependencies
+_fe_analyzer_shared  -         -           2.0.0       2.0.0     
+analyzer             0.35.4    0.35.4      0.39.5      0.39.5    
+...
+
+9  dependencies are constrained to versions that are older than a resolvable version.
+To update these dependencies, edit pubspec.yaml.
+```
+
+<span style="color:red">[PENDING: Is there a better example we should use?]</span>
+
+The preceding output shows that all regular dependencies are up to date,
+but one [dev dependency][] (`build_web_compilers`) is out of date.
+Several [transitive dependencies][] are also out of date.
+
+[dev dependency]: /tools/pub/dependencies#dev-dependencies
+[transitive dependencies]: /tools/pub/glossary#transitive-dependency
+
+
+To fix the dev dependency version,
+the easiest solution is to change the `build_web_compilers` entry
+in `pubspec.yaml` from `^1.0.0` to `^2.10.0`.
+For help on specifying versions, see the
+[version constraints documentation][constraints].
+
+[constraints]: /tools/pub/dependencies#version-constraints
+
+<span style="color:red">[PENDING: Improve the formatting of the following diff]</span>
+
+```diff
+-  build_web_compilers: ^1.0.0
++  build_web_compilers: ^2.10.0
+```
+
+{{site.alert.tip}}
+  To see what changed in the new version, you can look at the package's changelog,
+  which you can find in the package page on
+  [pub.dev.]({{site.pub}})
+  For an example, see the
+  [build_web_compilers changelog.]({{site.pub-pkg}}/build_web_compilers#-changelog-tab-)
+{{site.alert.end}}
+
+After editing `pubspec.yaml`, run `pub upgrade` to
+update the `pubspec.lock` file.
+You can then run `pub outdated` to confirm that
+you've made all necessary changes.
+
+```terminal
+$ pub upgrade
+...
+$ pub outdated
+Package     Current  Upgradable  Resolvable  Latest  
+dependencies: all up-to-date
+
+dev_dependencies: all up-to-date
+
+transitive dependencies
+googleapis  0.53.0   0.53.0      0.53.0      0.54.0  
+$
+```
+<span style="color:red">[PENDING: file bug about "up-to-date" (shouldn't be hyphenated unless it comes before a noun)]
+
+{{site.alert.important}}
+  Test your code to verify that it still works as expected
+  after updating the packages.
+{{site.alert.end}}
+
+If any transitive dependencies are out of date,
```

Comment:
I've found it easier to simply edit pubspec yaml to move the dependency that isn't resolvable, and then looking at the error output from pub get.

Reply:
Transitive dependencies aren't in pubspec.yaml... Do you mean remove that dependency from pubspec.lock?

Reply:
I don't think you can ever have any outdated transitive dependencies unless you either have outdated main dependencies, dev_dependencies, or both.

To figure out why I can't get to the most recent main or dev dep (listed in Latest), I edit pubspec.yaml to pin to that exact version, run pub get, and look at the resolution error.

Reply:
Couldn't you have a regular or dev dependency that, in its latest version, depends on something that's out-of-date?

Reply:
> I don't think you can ever have any outdated transitive dependencies unless you either have outdated main dependencies, dev_dependencies, or both.

<details>
<summary>You absolutely can :)</summary>
If you depend on `foo` and `foo` depends on `retry` version `^2.0.0`, then you won't get `retry` version `3.0.0` (despite the fact that it's only a transitive dependency).
</details>

Reply:
So... is `pub deps` the best way to figure out which package's [direct or transitive] dependencies are causing the problem?

Reply:
Yes, `pub deps -s list` is good for this. Take the example I created in https://github.com/kwalrath/po-example/pull/1. After fixing the first two issues (args to 1.6.0, http to 0.12.0) what remains is path stuck on 1.6.2 with 1.7.0 being the latest. Here's is what the output of deps is:

```
dependencies:
- args 1.6.0
- http 0.12.1
  - http_parser >=0.0.1 <4.0.0
  - path >=0.9.0 <2.0.0
  - pedantic ^1.0.0
- path 1.6.2
- terminal_tools 0.0.0
  - path 1.6.2
  - meta 1.1.6

dev dependencies:
[...]

transitive dependencies:
[...]
```

From this we can see path is allowed to be 1.6.2 because terminal_tools allows just that single version, and similar for meta.

Another option is to update `pubspec.yaml` to temporarily have that dep directly:

```
dependencies:
  args: ^1.4.4
  http: ^0.12.0
  # Changed this to 1.7.0 to trigger error:
  path: ^1.7.0
  terminal_tools:
    path: ./pkg/terminal_tools/

```

The resolution will tell you what's going on:

```
mit-macbookpro4:po-example mit$ ~/sdk/dart-beta/dart-sdk/bin/pub get
Resolving dependencies...
Because every version of terminal_tools from path depends on path 1.6.2 and po_example depends on path ^1.7.0,
  terminal_tools from path is forbidden.
So, because po_example depends on terminal_tools from path, version solving failed.
```

For the transitive that's a bit harder; you need to temporarily add the transitive dep as a direct:
```
dependencies:
  args: ^1.4.4
  http: ^0.12.0
  path: ^1.5.0
  terminal_tools:
    path: ./pkg/terminal_tools/
  # begin temp dep to trigger error
  meta: 1.1.8
  # end temp dep
```

The resolution will tell you what's going on:

```
$ ~/sdk/dart-beta/dart-sdk/bin/pub get
Resolving dependencies...
Because po_example depends on terminal_tools from path which depends on meta 1.1.6, meta 1.1.6 is required.
So, because po_example depends on meta 1.1.8, version solving failed.
```